Handle mysql YEAR column types (Fixes #13203) [ci drivers]

dpsutton opened the pull request:
- year columns are very small in the db. The tricky part is how the
driver hands them back: either as shorts or as java.sql.Dates. Our
standard trick of reading them right into the java.time.LocalDate
doesn't work and we instead get the java.sql.Date (which the driver
converts from the short internal representation) or the short if
`yearIsDateType` is configured to false.

It seems to be an open question if Metabase should always consume this
as a Date. That seems most likely but I have no idea how often this
comes up in practice.

Fixes https://github.com/metabase/metabase/issues/13203

dpsutton committed:
Handle mysql YEAR column types (Fixes #13203) [ci drivers]

- year columns are very small in the db. The tricky part is how the
driver hands them back: either as shorts or as java.sql.Dates. Our
standard trick of reading them right into the java.time.LocalDate
doesn't work and we instead get the java.sql.Date (which the driver
converts from the short internal representation) or the short if
`yearIsDateType` is configured to false.

It seems to be an open question if Metabase should always consume this
as a Date. That seems most likely but I have no idea how often this
comes up in practice.

dpsutton commented:
Should be straight forward fix. Open to comments about test location. Should it go into the fingerprint test file? I figured its best here as it relies on specific mysql features (YEAR) but open to strong opinions. Also, if anyone sees a way to cast without going to java.sql.Date in the driver I'm open but I didn't see any good way. The way we're doing it before (`(.getObject rs i java.time.LocalDate)` blows up here trying with

dpsutton committed:
Add typehints [ci drivers]

dpsutton commented:
Should be straight forward fix. Open to comments about test location. Should it go into the fingerprint test file? I figured its best here as it relies on specific mysql features (YEAR) but open to strong opinions.

Also, in:

```clojure
(defmethod sql-jdbc.execute/read-column-thunk [:mysql Types/DATE]
  [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
  (let [parent-thunk ((get-method sql-jdbc.execute/read-column-thunk [:sql-jdbc Types/DATE]) driver rs rsmeta i)]
    (fn read-time-thunk []
      ;; YEAR type is stored as smallint and exposed in the driver as either java.sql.Date or a short, depending on
      ;; the mysql configuration option `yearIsDateType`. The parent thunk will try (.getObject rs i
      ;; java.time.LocalDate) but this errors with "Cannot read LocalDate using a Types.SMALLINT
      ;; field". https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-type-conversions.html. (NOTE: the
      ;; documentation mentions the return type is java.sql.Short but I believe this is a typo for java.lang.Short:
      ;; https://bugs.mysql.com/bug.php?id=35115)
      (if (= "YEAR" (.getColumnTypeName rsmeta i))
        (let [x (.getObject rs i)]
          (condp instance? x
            java.sql.Date (.toLocalDate ^java.sql.Date x)
            java.lang.Short (long x)
            :else x))
        (parent-thunk)))))
```

is there a difference when the column type name is checked? Should that be checked and one of two thunks returned or is the current way, checking the metadata later sufficient? I'm a little worried about lifetimes and connections so I may change that.

dpsutton committed:
Consume metadata immediately when checking for YEAR mysql column

don't want to access the metadata outside of a `jdbc/with-db-metadata`

sbelak commented on src/metabase/driver/mysql.clj:
```diff
@@ -368,6 +368,24 @@
+          (condp instance? x
+            java.sql.Date (.toLocalDate ^java.sql.Date x)
+            java.lang.Short (long x)
+            :else x)))
```
when does this else come into play?

dpsutton replied:
that else handled when the column was null but i've changed it to make it more explicit.

sbelak reviewed (commented):
I think it's cleaner if we cast to some form of date, as it preserves semantic information. What kind of type do we assign these fields in our parlance?  

The test location is fine, since it's a db-specific thing.

I'd yank the col type and return one or the other thunk if nothing else for efficiency's sake.

dpsutton committed:
Clean up logic for YEAR column types [ci drivers]

the `:else` is actually bad syntax. And handle explicitly the nil case
and the two types it can be from getObject

dpsutton commented:
> I think it's cleaner if we cast to some form of date, as it preserves semantic information. What kind of type do we assign these fields in our parlance?

I'm gonna figure out how to set this config value so i can see what kind of shorts come back. I think they might be offsets from 1970 but not totally sure. Not sure what you mean what type we assign besides java.time.LocalDate.

> The test location is fine, since it's a db-specific thing.

👍 


> I'd yank the col type and return one or the other thunk if nothing else for efficiency's sake.

Done

sbelak commented:
> I'm gonna figure out how to set this config value so i can see what kind of shorts come back. I think they might be offsets from 1970 but not totally sure. Not sure what you mean what type we assign besides java.time.LocalDate.

I meant do we mark it as :type/Temporal or :type/Number

dpsutton committed:
Mark YEAR as :type/Date, make tests more reliable

column ordering was messing up testing, so ignore the PK and just
select the YEAR column

dpsutton commented:
@sbelak i think we're ready for approval unless you have some thoughts on `:type/Temporal` vs the one i chose for this `:type/Date`.

sbelak commented:
Temporal is the more generic. Since we cast to Date, I think it can be `:type/Date`.

sbelak commented on src/metabase/driver/mysql.clj:
```diff
@@ -368,6 +368,16 @@
+  [driver ^ResultSet rs ^ResultSetMetaData rsmeta ^Integer i]
+  (if (= "YEAR" (.getColumnTypeName rsmeta i))
+    (fn read-time-thunk []
+      (let [x (.getObject rs i)]
```
`when-let` ?

sbelak commented on src/metabase/driver/mysql.clj:
```diff
@@ -368,6 +368,16 @@
+    (fn read-time-thunk []
+      (let [x (.getObject rs i)]
+        (when x
+          (.toLocalDate ^java.sql.Date x))))
```
so it turns out we don't have to even handle a number here?

dpsutton committed:
Use `when-let` when checking YEAR column [ci drivers]

sbelak commented:
Codecov is complaining because Codecove runs only for H2 driver and this is all mysql-specific. The Book of Cam says: thou shall find something that's not covered, and add unrelated tests

dpsutton committed:
Add tests for `table-rows-sample`

really just to appease the Book of Cam® for test coverage

camsaul commented:
Please don't merge stuff in unless it's green. MySQL/MariaDB tests are failing now on `release-0.36.x`

camsaul commented:
@sbelak @dpsutton 

Going forward, for MySQL or Postgres driver code that Cloverage doesn't test we can go ahead and merge anyway even if CodeCov is red. This really only affects MySQL and Postgres since the other drivers are in other directories and are already excluded by Cloverage. I don't think it makes sense to have tests in completely unrelated areas be a requirement before merging any MySQL/Postgres fixes.

The goal of the CodeCov stuff is to make sure any new code or code we touch has a certain amount of coverage, not to force us to write tests in unrelated areas just so we can merge a PR. Sometimes refactors like moving code around will trigger a CodeCov failure if it didn't have adequate coverage in the first place. So in that case you might have to go write some tests to get things green. That's what I meant by finding something that's not covered [by the code you moved/refactored/etc.] and writing tests for it. 

I think we can have CodeCov exclude `metabase.driver.mysql` and `metabase.driver.postgres`, I'll have to take a look at that
